Sources/NIO/ConvenienceOptionSupport.swift

```diff
@@ -0,0 +1,178 @@
+    public static let allowImmediateLocalEndpointAddressReuse = NIOTCPShorthandOption(.reuseAddr)
+    
+    /// The user will manually call `Channel.read` once all the data is read from the transport.
+    public static let disableAutoRead = NIOTCPShorthandOption(.disableAutoRead)
+    
+    /// Allows users to configure whether the `Channel` will close itself when its remote
+    /// peer shuts down its send stream, or whether it will remain open. If set to `false` (the default), the `Channel`
+    /// will be closed automatically if the remote peer shuts down its send stream. If set to true, the `Channel` will
+    /// not be closed: instead, a `ChannelEvent.inboundClosed` user event will be sent on the `ChannelPipeline`,
+    /// and no more data will be received.
+    public static let allowRemoteHalfClosure =
+        NIOTCPShorthandOption(.allowRemoteHalfClosure)
+}
+
+/// A set of `NIOTCPShorthandOption`s
+public struct NIOTCPShorthandOptions : ExpressibleByArrayLiteral, Hashable {
+    var allowImmediateLocalEndpointAddressReuse = false
+    var disableAutoRead = false
+    var allowRemoteHalfClosure = false
+    
+    /// Construct from an array literal.
+    @inlinable
+    public init(arrayLiteral elements: NIOTCPShorthandOption...) {
+        for element in elements {
+            add(element)
+        }
+    }
+    
+    @usableFromInline
+    mutating func add(_ element: NIOTCPShorthandOption) {
+        switch element.data {
+        case .reuseAddr:
+            self.allowImmediateLocalEndpointAddressReuse = true
+        case .allowRemoteHalfClosure:
+            self.allowRemoteHalfClosure = true
+        case .disableAutoRead:
+            self.disableAutoRead = true
+        }
+    }
+    
+    /// Caller is consuming the knowledge that allowImmediateLocalEndpointAddressReuse was set or not.
+    /// The setting will nolonger be set after this call.
+    /// - Returns: If allowImmediateLocalEndpointAddressReuse was set.
+    public mutating func consumeAllowImmediateLocalEndpointAddressReuse() -> NIOOptionValue<Void> {
+        defer {
+            self.allowImmediateLocalEndpointAddressReuse = false
+        }
+        return NIOOptionValue<Void>(flag: self.allowImmediateLocalEndpointAddressReuse)
+    }
+    
+    /// Caller is consuming the knowledge that disableAutoRead was set or not.
+    /// The setting will nolonger be set after this call.
+    /// - Returns: If disableAutoRead was set.
+    public mutating func consumeDisableAutoRead() -> NIOOptionValue<Void> {
+        defer {
+            self.disableAutoRead = false
+        }
+        return NIOOptionValue<Void>(flag: self.disableAutoRead)
+    }
+    
+    /// Caller is consuming the knowledge that allowRemoteHalfClosure was set or not.
+    /// The setting will nolonger be set after this call.
+    /// - Returns: If allowRemoteHalfClosure was set.
+    public mutating func consumeAllowRemoteHalfClosure() -> NIOOptionValue<Void> {
```

Comment:
Are these three consume functions necessary? The fallback just directly looks at the stored properties on the struct: why shouldn't other use-cases do the same?

Reply:
They're not strictly necessary but they do push the user towards doing the right thing of removing them from the set when applying.